datumaro/plugins/synthetic_images_plugin/image_generator.py

```diff
@@ -0,0 +1,174 @@
+import numpy as np
+
+from datumaro.components.cli_plugin import CliPlugin
+from datumaro.components.errors import SourceExistsError
+
+from .utils import IFSFunction, augment, colorize, download_colorization_model
+
+
+class ImageGenerator(CliPlugin):
+    """
+    ImageGenerator generates a 3-channel synthetic dataset with provided shapes.
+    """
+
+    @classmethod
+    def build_cmdline_parser(cls, **kwargs):
+        parser = super().build_cmdline_parser(**kwargs)
+        parser.add_argument('-k', '--count', type=int, required=True,
+            help="Number of images to generate")
+        parser.add_argument('-o', '--output-dir', type=os.path.abspath, required=True,
+            help="Path to the directory where dataset are saved")
+        parser.add_argument('--shape', nargs='+', type=int,
+            help="Image shape: height and width separated by a space, "
+                 "for example --shape 256 224")
+        parser.add_argument('--overwrite', action='store_true',
+            help="Overwrite existing files in the save directory")
+
+        return parser
+
+    def __init__(self, count: int, output_dir: str,
+                 shape: List[int], overwrite: bool = False):
+        self._count = count
+        self._output_dir = output_dir
+        assert len(shape) == 2
+        self._height, self._width = shape
+
+        if not os.path.exists(output_dir):
+            os.mkdir(output_dir)
+        elif os.path.isdir(output_dir) and os.listdir(output_dir):
+            if overwrite:
+                rmtree(output_dir)
+                os.mkdir(output_dir)
+            else:
+                raise SourceExistsError(output_dir)
+
+        np.random.seed(0)
+        self._cpu_count = min(os.cpu_count(), self._count)
+        self._weights = np.array([
+            0.2, 1, 1, 1, 1, 1,
+            0.6, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1,
+            1.4, 1, 1, 1, 1, 1,
+            1.8, 1, 1, 1, 1, 1,
+            1, 0.2, 1, 1, 1, 1,
+            1, 0.6, 1, 1, 1, 1,
+            1, 1.4, 1, 1, 1, 1,
+            1, 1.8, 1, 1, 1, 1,
+            1, 1, 0.2, 1, 1, 1,
+            1, 1, 0.6, 1, 1, 1,
+            1, 1, 1.4, 1, 1, 1,
+            1, 1, 1.8, 1, 1, 1,
+            1, 1, 1, 0.2, 1, 1,
+            1, 1, 1, 0.6, 1, 1,
+            1, 1, 1, 1.4, 1, 1,
+            1, 1, 1, 1.8, 1, 1,
+            1, 1, 1, 1, 0.2, 1,
+            1, 1, 1, 1, 0.6, 1,
+            1, 1, 1, 1, 1.4, 1,
+            1, 1, 1, 1, 1.8, 1,
+            1, 1, 1, 1, 1, 0.2,
+            1, 1, 1, 1, 1, 0.6,
+            1, 1, 1, 1, 1, 1.4,
+            1, 1, 1, 1, 1, 1.8,
+        ]).reshape(-1, 6)
+        self._threshold = 0.2
+        self._iterations = 200000
+        self._num_of_points = None
+        self._instances = None
+        self._categories = None
+
+    def generate_dataset(self) -> None:
+        path = os.path.dirname(os.path.abspath(__file__))
+        download_colorization_model(path)
+        self._initialize_params()
```

Comment:
Why is this called here? If it does initialization, why isn't it a part of the constructor?

Reply:
Moved to constructor